packages/utreexo/src/stump/proof.cairo

```
@@ -27,3 +30,150 @@ impl UtreexoBatchProofDisplay of Display<UtreexoBatchProof> {
+                    next_computed_pos = a;
+                    next_computed = b;
+                }
+                res
+            } else {
+                // Out of nodes, terminating here.
+                break;
+            };
+
+            // If we are beyond current row, level up.
+            while pos >= row_len_acc + row_len {
+                row_len_acc += row_len;
+                row_len /= 2;
+                actual_row_len /= 2;
+            };
+
+            // If row length is odd and we are at the edge this is a root.
+            if pos == row_len_acc + actual_row_len - 1 && actual_row_len % 2 == 1 {
+                calculated_root_hashes.append(node);
+                row_len_acc += row_len;
+                row_len /= 2;
+                actual_row_len /= 2;
+                continue;
+            };
+
+            let parent_node = if (pos - row_len_acc) % 2 == 0 {
+                // Right sibling can be both leaf/computed or proof.
+                let right_sibling = if next_leaf_pos == pos + 1 {
+                    let res = next_leaf;
+                    if leaf_nodes.is_empty() {
+                        next_leaf_pos = Bounded::<u64>::MAX;
+                    } else {
+                        let (a, b) = leaf_nodes.pop_front().unwrap();
+                        next_leaf_pos = a;
+                        next_leaf = b;
+                    }
+                    res
+                } else if next_computed_pos == pos + 1 {
+                    let res = next_computed;
+                    if computed_nodes.is_empty() {
+                        next_computed_pos = Bounded::<u64>::MAX;
+                    } else {
+                        let (a, b) = computed_nodes.pop_front().unwrap();
+                        next_computed_pos = a;
+                        next_computed = b;
+                    }
+                    res
+                } else {
+                    sibling_nodes.pop_front().unwrap()
```

Let's return an error if proof is empty (via `inner_result`)